Add language parameter to Text objects

QuLogic opened the pull request:
## PR summary

Along with #29695, this *language* parameter is an additional setting that may affect text layout. As with the other PR, this is not complete, but rather something to get the API decided upon (and likely will be rebased on raqm PR after it's done.)

There are two parts to the API:
1. the `language` parameter attached to `Text`, and any other API that wraps it (i.e., `Axes.text` or `Axes.annotate`)
2. the `language` parameter in backend's `Renderer.draw_text`

At the moment, for both, I have set these to `str | list[tuple[str, int, int]]` where the latter denotes a list of (language, start, end)-tuples. However, this was before I found out about [the font feature machinery](https://github.com/matplotlib/matplotlib/pull/29695#issuecomment-2692176765), and it is possible that we may wish to use that syntax instead, at least for API part 1. For part 2, it's a little nicer to have the explicit types, just for passing to the FT2Font extension, but this may not be relevant to other implementations.

## PR checklist
- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

anntzer commented:
re: adding language to draw_text: the general discussion about extensibility of renderer API applies, i.e.
- if you're going to add language as a parameter to draw_text then it should be at least keyword-only.
- it's likely that the more maintainable API would likely to just embed language as part of fontproperties (it's not really different from e.g. font features, I'd think) and tell the backend to read that info.
- for draw_text we have actually more or less given up on having a "clean" renderer API because we pass the whole Text instance to draw_text (per #1081) anyways so perhaps backends that care could also just read that info off the Text instance.  It feels a bit dirty, but heh...

anntzer commented:
Side point regarding allowing `list[tuple[str, int, int]]` as language setting: this may be a problem if one uses `somelanguage[idx:]` to set the language from an index up to the end of the string (whose length could technically be modified later, even though it's not clear this necessarily makes much practical sense).  Likely you need `list[tuple[str, int, int | None]]` (and then you may possibly support None (=0) for the start as well).

QuLogic commented:
In the call a couple weeks ago, we decided to drop the new `language` parameter to `draw_text`, because that already had a `mtext` parameter which contained the whole `Text` object. This would avoid any signature introspection.

However, one followup is that `Renderer.get_text_width_height_descent` would _also_ depend on font features/language settings, and really should  get these passed through somehow as well.

QuLogic commented on lib/matplotlib/_text_helpers.py:
```diff
@@ -65,7 +65,7 @@ def layout(string, font, *, kern_mode=Kerning.DEFAULT):
     x = 0
     prev_glyph_idx = None
-    char_to_font = font._get_fontmap(string)
+    char_to_font = font._get_fontmap(string)  # TODO: Pass in language.
```
Note, this function is getting rewritten with libraqm, so this is just a note for later.

anntzer commented on lib/matplotlib/_text_helpers.py:
```diff
@@ -43,7 +43,7 @@ def warn_on_missing_glyph(codepoint, fontnames):
 
 
-def layout(string, font, *, kern_mode=Kerning.DEFAULT):
+def layout(string, font, language, *, kern_mode=Kerning.DEFAULT):
```
These params should be kwonly? (throughout most of the PR, I think)

QuLogic commented on src/ft2font_wrapper.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <cstddef>
```
Accidental; `clangd` likes to add these itself and sometimes I miss it.

QuLogic commented:
I've added a `text.language` rcParam as discussed on the call last week.

QuLogic commented:
Also added a What's new note, though as with the other PR, it won't show the difference until `libraqm` is added.

QuLogic commented on doc/users/next_whats_new/text_language.rst:
```diff
@@ -0,0 +1,37 @@
+    fig.text(0.5, 0.3, f'\\U{ord(char):08x}', fontsize=40, horizontalalignment='center')
+    fig.text(0, 0.1, f'English: {char}', fontsize=40, language='en')
+    fig.text(1, 0.1, f'Inari Sámi: {char}', fontsize=40, language='smn',
+             horizontalalignment='right')
```
I think we're not ready to expose that just yet, so it's not been documented.

QuLogic commented on lib/matplotlib/tests/test_text.py:
```diff
@@ -1190,3 +1190,19 @@ def test_ytick_rotation_mode():
+    with pytest.raises(TypeError, match='must be list of tuple'):
+        Text(0, 0, 'foo', language=[1, 2, 3])
+    with pytest.raises(TypeError, match='must be list of tuple'):
+        Text(0, 0, 'foo', language=[(1, 2)])
```
Yes, but the tuple is the wrong length (I just didn't write out the whole message.)

QuLogic commented on lib/matplotlib/text.py:
```diff
@@ -1422,6 +1423,41 @@ def _va_for_angle(self, angle):
+        _api.check_isinstance((list, str, None), language=language)
+        language = mpl._val_or_rc(language, 'text.language')
+
+        if isinstance(language, list):
```
No, that would not make sense, but it's understandable that it's not clear since it's purposely undocumented. If you provide multiple languages, then they would need to specify a range over which they apply.

QuLogic commented on lib/matplotlib/text.py:
```diff
@@ -1422,6 +1423,41 @@ def _va_for_angle(self, angle):
+        _api.check_isinstance((list, str, None), language=language)
+        language = mpl._val_or_rc(language, 'text.language')
+
+        if isinstance(language, list):
```
I'll allow `tuple` as well, though since there's no fixed size, it won't be allowed by type hints.

QuLogic commented:
Rebased on the text-overhaul branch and the failures should be fixed with the preloading from #30231.

tacaswell commented on lib/matplotlib/text.py:
```diff
@@ -1422,6 +1423,41 @@ def _va_for_angle(self, angle):
+
+        Parameters
+        ----------
+        language : str or None
```
```suggestion
        language : str or list or tuple or None
```

tacaswell replied:
I was about the ask the same question looking at the tests! 

Marking this as resolved.

QuLogic commented:
Since #30000 is almost merged, I've rebased on top of that, and added the libraqm implementation, so the docs now correctly show the change and there are working tests.

QuLogic commented on src/ft2font.cpp:
```diff
@@ -361,6 +363,17 @@ void FT2Font::set_text(
 
+    if (languages.has_value()) {
+        for (auto & [lang_str, start, end] : languages.value()) {
+            if (!raqm_set_language(rq, lang_str.c_str(), start, end)) {
```
This function actually takes a length, not an end location; I'm not sure which one we should accept, but I'll have to update this to the correct naming (the test is setup to match what libraqm says, but the docs say the opposite.)

anntzer replied:
start-stop looks normal as API to me?

QuLogic commented:
Also extended the tests a bit to ensure that the results are as expected and fixed the fallback to the `rcParam` as well.

tacaswell commented on doc/release/next_whats_new/text_language.rst:
```diff
@@ -0,0 +1,37 @@
+
+    char = '\U0000014a'
+    fig.text(0.5, 0.3, f'\\U{ord(char):08x}', fontsize=40, horizontalalignment='center')
+    fig.text(0, 0.1, f'English: {char}', fontsize=40, language='en')
+    fig.text(1, 0.1, f'Inari Sámi: {char}', fontsize=40, language='smn',
```
Super picky change to match the example ordering to the text ordering:

```suggestion
    fig.text(1, 0.1, f'Inari Sámi: {char}', fontsize=40, language='smn')
    fig.text(0, 0.1, f'English: {char}', fontsize=40, language='en',
```

tacaswell commented:
It looks like the appveyor tests hung.
